test/conformance/channel_tracing_test.go

```diff
@@ -0,0 +1,173 @@
+// +build e2e
+
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conformance
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"knative.dev/eventing/test/base/resources"
+	"knative.dev/eventing/test/common"
+	"knative.dev/pkg/test/zipkin"
+)
+
+func TestChannelTracing(t *testing.T) {
+	testCases := map[string]struct {
+		incomingTraceId bool
+		istio           bool
+	}{
+		"no incoming trace id": {},
+		"includes incoming trace id": {
+			incomingTraceId: true,
+		}, /*
+			"caller has istio": {
+				istio: true,
+			},*/
+	}
+
+	for n, tc := range testCases {
+		loggerPodName := "logger"
+		t.Run(n, func(t *testing.T) {
+			channelTestRunner.RunTests(t, common.FeatureBasic, func(st *testing.T, channel string) {
+				// Don't accidentally use t, use st instead. To ensure this, shadow 't' to some a
+				// useless type.
+				t := struct{}{}
+				_ = fmt.Sprintf("%s", t)
+
+				client := common.Setup(st, true)
+				defer common.TearDown(client)
+
+				// Do NOT call zipkin.CleanupZipkinTracingSetup. That will be called exactly once in
+				// TestMain.
+				zipkin.SetupZipkinTracing(client.Kube.Kube, st.Logf)
+
+				// TODO This should really be upsert.
+				err := client.Kube.UpdateConfigMap("knative-eventing", "config-tracing", map[string]string{
+					"backend":         "zipkin",
+					"zipkin-endpoint": "http://zipkin.istio-system.svc.cluster.local:9411/api/v2/spans",
+					"debug":           "true",
+					"sample-rate":     "1.0",
+				})
+				if err != nil {
+					st.Fatalf("Unable to set the ConfigMap: %v", err)
+				}
+
+				mustContain := setupChannelTracing(st, channel, client, loggerPodName, tc.incomingTraceId)
+				assertLogContents(st, client, loggerPodName, mustContain)
+
+				traceID := getTraceID(st, client, loggerPodName)
+
+				// We expect the following spans:
+				// 1. Sending pod sends event to Channel (only if the sending pod generates a span).
+				// 2. Channel receives event from sending pod.
+				// 3. Channel sends event to logging pod.
+				// 4. Logging pod receives event from Channel.
+				// So we expect 4 spans if the sending pod is generating a span, 3 if not.
+				expectedNumSpans := 3
+				if tc.incomingTraceId {
+					expectedNumSpans = 4
+				}
+				trace, err := zipkin.JSONTrace(traceID, expectedNumSpans, 60*time.Second)
```

Comment:
The test will fail if I run locally on my MAC, I haven't figured out the root cause, but removing the check in https://github.com/knative/pkg/blob/5cdf30f51ef152232da0e56ef39d35de6c057f3d/test/zipkin/util.go#L167 can solve it.

Reply:
Genius! I fixed this by removing it in knative/pkg#693.